linker/shared/src/main/scala/org/scalajs/linker/frontend/modulesplitter/MaxModuleAnalyzer.scala

```diff
@@ -201,19 +164,52 @@ private object MaxModuleAnalyzer {
       }
     }
 
-    def tags(): (Set[ModuleID], Set[ClassName]) = {
-      /* Remove dynamic paths to class that are also reached by a public module.
-       * However, only do this if there are other tags as well. Otherwise, this
-       * class will end up in a public module, but the dynamically loaded module
-       * will try to import it (but importing public modules is forbidden).
-       */
-      if (direct.size > 1 || direct != dynamic.keySet) {
-        direct.foreach(dynamic.remove(_))
+    def moduleID(internalModuleIDPrefix: String): ModuleID = {
+      if (direct.size == 1 && dynamic.isEmpty) {
+        /* Class is only used by a single public module. Put it there.
+         *
+         * Note that we must not do this if there are any dynamic modules
+         * requiring this class. Otherwise, the dynamically loaded module
+         * will try to import the public module (but importing public modules is
+         * forbidden).
+         */
+        direct.head
+      } else {
+        /* Class is used by multiple public modules and/or dynamic edges.
+         * Create a module ID grouping it with other classes that have the same
+         * dependees.
+         */
+        val digestBuilder = new SHA1.DigestBuilder
+
+        // Public modules using this.
+        for (id <- direct.toList.sortBy(_.id))
+          digestBuilder.update(id.id.getBytes(StandardCharsets.UTF_8))
+
+        // Dynamic modules using this.
+        for (className <- dynamicEnds)
+          digestBuilder.updateUTF8String(className.encoded)
+
+        // Build a hex string of the hash with the right prefix.
+        val id = new StringBuilder(internalModuleIDPrefix)
```

Comment:
This a Scala `StringBuilder`, which is an unnecessarily expensive wrapper over `java.lang.StringBuilder`. Consider using the latter directly:
```suggestion
        val id = new java.lang.StringBuilder(internalModuleIDPrefix)
```